waspc/packages/wasp-config/src/appSpec.ts

```diff
@@ -0,0 +1,210 @@
+/** This module is a mirror implementation of AppSpec Decls in TypeScript.
+ * The original implemention is in Haskell (waspc).
+ *
+ * IMPORTANT: Do not change this file without updating the AppSpec in waspc.
+ */
+
+export type Decl =
+  | { declType: 'App'; declName: string; declValue: App }
+  | { declType: 'Page'; declName: string; declValue: Page }
+  | { declType: 'Route'; declName: string; declValue: Route }
+  | { declType: 'Query'; declName: string; declValue: Query }
+  | { declType: 'Action'; declName: string; declValue: Action }
+  | { declType: 'App'; declName: string; declValue: App }
+  | { declType: 'Job'; declName: string; declValue: Job }
+  | { declType: 'Api'; declName: string; declValue: Api }
+  | { declType: 'ApiNamespace'; declName: string; declValue: ApiNamespace }
+  | { declType: 'Crud'; declName: string; declValue: Crud }
```

Comment:
This looks a little repetitive, but removing the repetition with a complicated generic isn't worth it IMO.